src/include/simeng/PerceptronPredictor.hh

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include <deque>
+#include <map>
+#include <vector>
+
+#include "simeng/BranchPredictor.hh"
+#include "simeng/config/SimInfo.hh"
+
+namespace simeng {
+
+/** A Perceptron branch predictor implementing the branch predictor described in
+ * Jimenez and Lin ("Dynamic branch prediction with perceptrons", IEEE High-
+ * Performance Computer Architecture Symposium Proceedings (2001), 197-206 --
+ * https://www.cs.utexas.edu/~lin/papers/hpca01.pdf).
+ * The following predictors have been included:
+ *
+ * - Static predictor based on pre-allocated branch type.
+ *
+ * - A Branch Target Buffer (BTB) with a local and global indexing scheme and a
+ * perceptron.
+ *
+ * - A Return Address Stack (RAS) is also in use.
+ */
+
+class PerceptronPredictor : public BranchPredictor {
+ public:
+  /** Initialise predictor models. */
+  PerceptronPredictor(ryml::ConstNodeRef config = config::SimInfo::getConfig());
+  ~PerceptronPredictor();
+
+  /** Generate a branch prediction for the supplied instruction address, a
+   * branch type, and a known branch offset; defaults to 0 meaning offset is not
+   * known. Returns a branch direction and branch target address. */
+  BranchPrediction predict(uint64_t address, BranchType type,
+                           int64_t knownOffset = 0) override;
+
+  /** Updates appropriate predictor model objects based on the address and
+   * outcome of the branch instruction. */
+  void update(uint64_t address, bool taken, uint64_t targetAddress,
+              BranchType type) override;
+
+  /** Provides RAS rewinding behaviour. */
+  void flush(uint64_t address) override;
+
+ private:
+  /** Returns the dot product of a perceptron and a history vector.  Used to
+   * determine a direction prediction */
+  int64_t getDotProduct(const std::vector<int8_t> perceptron, uint64_t history);
```

Comment:
Could `perceptron` be passed as `const std::vector<int8_t>& perceptron`

Reply:
yes, it could.  Changed.